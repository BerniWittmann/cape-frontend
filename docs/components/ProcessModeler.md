# ProcessModeler

A Component which allows the process itself to be edited. This is a wrapper for the bpmn-io library

## Props

<!-- @vuese:ProcessModeler:props:start -->
|Name|Description|Type|Required|Default|
|---|---|---|---|---|
|value|The value object containing the keys `xml`, `svg`, `extensionAreas`|`Object`|`true`|-|

<!-- @vuese:ProcessModeler:props:end -->


## Events

<!-- @vuese:ProcessModeler:events:start -->
|Event Name|Description|Parameters|
|---|---|---|
|input|Fired when the data changed|The new value object with `svg`, `xml`and `extensionAreas`|

<!-- @vuese:ProcessModeler:events:end -->


## Methods

<!-- @vuese:ProcessModeler:methods:start -->
|Method|Description|Parameters|
|---|---|---|
|validate|Validate the Process Model|-|
|reloadXML|Reload the XML data into the modeler|-|

<!-- @vuese:ProcessModeler:methods:end -->


