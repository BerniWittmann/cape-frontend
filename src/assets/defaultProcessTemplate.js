export default '<?xml version="1.0" encoding="UTF-8"?> ' +
'<bpmn:definitions xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:bpmn="http://www.omg.org/spec/BPMN/20100524/MODEL" xmlns:bpmndi="http://www.omg.org/spec/BPMN/20100524/DI" xmlns:dc="http://www.omg.org/spec/DD/20100524/DC" xmlns:di="http://www.omg.org/spec/DD/20100524/DI" id="Definitions_0pzq2xr" targetNamespace="http://bpmn.io/schema/bpmn"> ' +
'<bpmn:process id="Process_1" isExecutable="false"> ' +
'<bpmn:startEvent id="StartEvent_1"> ' +
'<bpmn:outgoing>SequenceFlow_1efzkxz</bpmn:outgoing> ' +
'</bpmn:startEvent> ' +
'<bpmn:intermediateThrowEvent id="IntermediateThrowEvent_12et4wi"> ' +
'<bpmn:incoming>SequenceFlow_1efzkxz</bpmn:incoming> ' +
'</bpmn:intermediateThrowEvent> ' +
'<bpmn:sequenceFlow id="SequenceFlow_1efzkxz" sourceRef="StartEvent_1" targetRef="IntermediateThrowEvent_12et4wi" /> ' +
'</bpmn:process> ' +
'<bpmndi:BPMNDiagram id="BPMNDiagram_1"> ' +
'<bpmndi:BPMNPlane id="BPMNPlane_1" bpmnElement="Process_1"> ' +
'<bpmndi:BPMNShape id="_BPMNShape_StartEvent_2" bpmnElement="StartEvent_1"> ' +
'<dc:Bounds x="173" y="102" width="36" height="36" /> ' +
'</bpmndi:BPMNShape> ' +
'<bpmndi:BPMNShape id="IntermediateThrowEvent_12et4wi_di" bpmnElement="IntermediateThrowEvent_12et4wi"> ' +
'<dc:Bounds x="259" y="102" width="36" height="36" /> ' +
'</bpmndi:BPMNShape> ' +
'<bpmndi:BPMNEdge id="SequenceFlow_1efzkxz_di" bpmnElement="SequenceFlow_1efzkxz"> ' +
'<di:waypoint x="209" y="120" /> ' +
'<di:waypoint x="259" y="120" /> ' +
'</bpmndi:BPMNEdge> ' +
'</bpmndi:BPMNPlane> ' +
'</bpmndi:BPMNDiagram>' +
'</bpmn:definitions>'
