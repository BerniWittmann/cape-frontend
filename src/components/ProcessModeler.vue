<template>
  <div class="modeler">
    <div id="canvas"></div>
  </div>
</template>

<script>
/* ============
 * Process Modeler Component
 * ============
 *
 * A component to model a process. This component is a wrapper for the bpmn-io library
 */

import Modeler from 'bpmn-js/lib/Modeler'
import defaultProcessTemplate from '@/assets/defaultProcessTemplate'

export default {
  props: {
    value: {
      type: Object,
      required: true,
      validator: function (value) {
        return value.hasOwnProperty('xml') && value.hasOwnProperty('svg')
      }
    }
  },

  data() {
    return {
      modeler: undefined,
      isLoaded: false
    }
  },

  methods: {
    getSVG(cb) {
      if (this.modeler) {
        return this.modeler.saveSVG((err, svg) => {
          if (err) return cb(err)

          this.value.svg = svg
          return cb(err, svg)
        })
      }
      return cb()
    },

    getXML(cb) {
      if (this.modeler) {
        return this.modeler.saveXML((err, xml) => {
          if (err) return cb(err)

          this.value.xml = xml
          return cb(err, xml)
        })
      }
      return cb()
    },

    updateValues(event) {
      this.getXML(() => {
        this.getSVG(() => {
          this.$emit('input', this.value)
        })
      })
    }
  },

  mounted() {
    this.modeler = new Modeler({ container: '#canvas' })

    const xml = this.value.xml ? this.value.xml : defaultProcessTemplate

    this.modeler.importXML(xml, (err) => {
      if (err) throw err

      const canvas = this.modeler.get('canvas')
      canvas.zoom('fit-viewport')

      this.isLoaded = true
    })

    this.modeler.on('element.changed', this.updateValues)
    this.modeler.on('commandStack.changed', this.updateValues)
  }
}

</script>

<style lang="scss">
.modeler {
  border-radius: 5px;
  padding: 10px;
  background-color: white;
  border: 1px solid #909399;
  margin-bottom: 20px;
}
#canvas {
  height: 550px;
}
</style>
