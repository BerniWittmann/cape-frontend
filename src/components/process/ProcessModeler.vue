<template>
  <el-card class="modeler">
    <div id="canvas"></div>
  </el-card>
</template>

<script>
/* ============
 * Process Modeler Component
 * ============
 *
 * A component to model a process. This component is a wrapper for the bpmn-io library
 */

import Modeler from 'bpmn-js/lib/Modeler'
import BPMNModules from '@/plugins/bpmn/modules'

import defaultProcessTemplate from '@/assets/defaultProcessTemplate'

export default {
  model: {
    prop: 'value',
    event: 'input'
  },

  props: {
    value: {
      type: Object,
      required: true
    }
  },

  data() {
    return {
      modeler: undefined,
      isLoaded: false
    }
  },

  methods: {
    getSVG(cb) {
      if (this.modeler) {
        return this.modeler.saveSVG((err, svg) => {
          if (err) return cb(err)

          this.value.svg = svg
          return cb(err, svg)
        })
      }
      return cb()
    },

    getXML(cb) {
      if (this.modeler) {
        return this.modeler.saveXML((err, xml) => {
          if (err) return cb(err)

          this.value.xml = xml
          return cb(err, xml)
        })
      }
      return cb()
    },

    updateValues(event) {
      this.getXML(() => {
        this.getSVG(() => {
          this.$emit('input', this.value)
        })
      })
    }
  },

  mounted() {
    this.modeler = new Modeler({ container: '#canvas', additionalModules: BPMNModules })

    const xml = this.value.xml ? this.value.xml : defaultProcessTemplate

    this.modeler.importXML(xml, (err) => {
      if (err) throw err

      const canvas = this.modeler.get('canvas')
      canvas.zoom('fit-viewport')

      this.isLoaded = true
    })

    this.modeler.on('element.changed', this.updateValues)
    this.modeler.on('commandStack.changed', this.updateValues)
  }
}

</script>

<style lang="scss">
.modeler {
  border-radius: 5px;
  background-color: white;
  margin-bottom: 20px;
}
#canvas {
  height: 530px;
}
</style>
