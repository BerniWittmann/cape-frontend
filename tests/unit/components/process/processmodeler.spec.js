import { shallowMount } from '@vue/test-utils'
import { i18n } from '../../setupPlugins'

import Modeler from 'bpmn-js/lib/Modeler'
import ProcessModeler from '@/components/process/ProcessModeler.vue'
import BPMNModules from '@/plugins/bpmn/modules'
import defaultProcessTemplate from '@/assets/defaultProcessTemplate'

jest.mock('bpmn-js/lib/Modeler', () => {
  return jest.fn().mockImplementation(() => {
    return {
      on: jest.fn(),
      importXML: jest.fn().mockImplementation((xml, cb) => {
        if (xml.includes('####')) return cb(new Error('Invalid XML'))
        return cb(undefined)
      }),
      saveXML: jest.fn().mockImplementation((cb) => {
        return cb(undefined, '<xml></xml>')
      }),
      saveSVG: jest.fn().mockImplementation((cb) => {
        return cb(undefined, '<svg></svg>')
      }),
      get: jest.fn().mockImplementation(() => {
        return { zoom: jest.fn() }
      })
    }
  })
})

describe('Components', () => {
  describe('ProcessModeler', () => {
    let propsData
    let cmp

    beforeEach(() => {
      propsData = {
        value: {
          xml: undefined,
          svg: undefined
        }
      }
      render()
    })

    function render() {
      cmp = shallowMount(ProcessModeler, {
        i18n,
        propsData
      })
    }

    it('renders', () => {
      expect(cmp.html()).toMatchSnapshot()
    })

    it('sets up the modeler', () => {
      expect(Modeler).toHaveBeenCalledWith({
        container: '#canvas',
        additionalModules: BPMNModules
      })
    })

    it('sets up the change event listeners', () => {
      expect(cmp.vm.modeler.on).toHaveBeenCalledWith('element.changed', expect.any(Function))
      expect(cmp.vm.modeler.on).toHaveBeenCalledWith('commandStack.changed', expect.any(Function))
    })
    it('updates the svg on change event', () => {
      const fn = cmp.vm.modeler.on.mock.calls[0][1]
      fn()
      expect(cmp.vm.value.svg).toEqual('<svg></svg>')
      expect(cmp.emitted('input')).toBeTruthy()
      expect(cmp.emitted('input')[0][0].svg).toEqual('<svg></svg>')
    })
    it('updates the xml on change event', () => {
      const fn = cmp.vm.modeler.on.mock.calls[0][1]
      fn()
      expect(cmp.vm.value.xml).toEqual('<xml></xml>')
      expect(cmp.emitted('input')).toBeTruthy()
      expect(cmp.emitted('input')[0][0].xml).toEqual('<xml></xml>')
    })
    it('handles errors in change event for svg', () => {
      cmp.vm.modeler.saveSVG = jest.fn().mockImplementation((cb) => {
        return cb(new Error('Test Error'))
      })
      const fn = cmp.vm.modeler.on.mock.calls[0][1]
      fn()
      expect(cmp.emitted('input')).toBeTruthy()
      expect(cmp.emitted('input')[0][0].svg).toEqual(undefined)
    })
    it('handles errors in change event for xml', () => {
      cmp.vm.modeler.saveXML = jest.fn().mockImplementation((cb) => {
        return cb(new Error('Test Error'))
      })
      const fn = cmp.vm.modeler.on.mock.calls[0][1]
      fn()
      expect(cmp.emitted('input')).toBeTruthy()
      expect(cmp.emitted('input')[0][0].xml).toEqual(undefined)
    })
    it('handles not setup modeler in change event for svg', () => {
      const fn = cmp.vm.modeler.on.mock.calls[0][1]
      cmp.vm.modeler = undefined
      fn()
      expect(cmp.emitted('input')).toBeTruthy()
      expect(cmp.emitted('input')[0][0].svg).toEqual(undefined)
    })
    it('handles not setup modeler in change event for xml', () => {
      const fn = cmp.vm.modeler.on.mock.calls[0][1]
      cmp.vm.modeler = undefined
      fn()
      expect(cmp.emitted('input')).toBeTruthy()
      expect(cmp.emitted('input')[0][0].xml).toEqual(undefined)
    })
    it('imports the defaultXMLTemplate on empty process', () => {
      propsData.value.xml = undefined
      render()

      expect(cmp.vm.modeler.importXML).toHaveBeenCalledWith(defaultProcessTemplate, expect.any(Function))
    })
    it('imports the correct xml', () => {
      propsData.value.xml = '<xml></xml>'
      render()

      expect(cmp.vm.modeler.importXML).toHaveBeenCalledWith('<xml></xml>', expect.any(Function))
    })
    it('handles error on xml import', () => {
      console.error = jest.fn()
      propsData.value.xml = '<xml>####</xml>'
      expect(() => {
        render()
      }).toThrow()
    })
  })
})
