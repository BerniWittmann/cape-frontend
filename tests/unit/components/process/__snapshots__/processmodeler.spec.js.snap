// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components ProcessModeler renders 1`] = `
<elcard-stub class="modeler">
  <div id="canvas"></div>
</elcard-stub>
`;
